### ISSUE TITLE
x32 port

### ISSUE THREAD

**Issue by piranna** [NodeOS/NodeOS#253]
Create a port x32 platform

**Comment by lite20** [comment]
Do we only have x86? What are your motives behind creating a x32 variant? (Out of curiosity).

**Comment by piranna** [comment]
Currently ia32 (x86) and x86_64 (amd64), I got the kernel to compile and
boot for arm but didn't get it to work correctly.

x32 generates 64 bits code the same way as x86_64 but using pointers and
ints of 32 bits, using less memory (similar to ia32), so you get the best
of both worlds. The only drawback it has is that although being 64 bits
process can manage only up to 4GB of ram, but I find this more than
convenient and in fact if we do the port I would sort of deprecate the
x86_64 one...

**Comment by lite20** [comment]
@piranna ah mkay. So is it 4 gigs of ram per node process? Would there be a way to escape into the 64bit space and use more ram? 

What would we have to accomplish in order to get a working x32? I'm assuming that ia32 and x32 aren't the same thing?

**Comment by lite20** [comment]
@piranna are you referring to the x32 ABI? Maybe we should mention "ABI" for clarity, because I a bit confused at first, and others probably will be also.

**Comment by lite20** [comment]
@piranna so I did some research to see where v8 stands in terms of x32 support and there's a **lot** of work that would go into porting to the x32 abi platform. Here's an extract from a thread discussing porting to the x32  abi with one of the v8 developers:

>>>> OK - do you plan to add that support? Just asking e.g. so I can take the right action as maintainer of v8 package in Gentoo. It's fine for me if you don't have such plans. :)
> 
>>> No. (For reasoning, see below.)
> Adding support is not just a matter of fixing a few lines. It would be an entire new platform port (albeit similar to the existing ia32/x64 ports); roughly 40K lines of code.
> 
>> Got it. For reference - would you accept patches adding such a port? Note that I'm rather not going to work on it atm, but someone else might be interested.
> 
> It's not just the initial creation of a new platform port that's a bunch of work, it's also the ongoing maintenance of it (i.e. making it keep up with the continuing development of V8). As long as the potential user base for x32 is as small as it currently is, it doesn't make sense for the team to spend resources on that. 
> 
> The obvious alternative is a third-party maintained port in a separate repository that tracks V8, as our MIPS port is currently doing it (https://github.com/paul99/v8m-rb). This model can work with or without upstreaming of the changes, but either way with a clear statement that the core V8 team does not commit to making sure the port works and is up to date.

The whole thread is here: https://groups.google.com/forum/#!topic/v8-users/c-_URSZqTq8
We can attempt to undertake the port as a third party and people may potentially join, or we could wait for the people using the platform to grow and be sufficient for the v8 team to do it themselves. As massive an undertaking as this might be, the performance gains shouldn't be overlooked. Interesting suggestion!

**Comment by piranna** [comment]
> So is it 4 gigs of ram per node process?

Yes, the same as in ia32/x86.


> Would there be a way to escape into the 64bit space and use more ram?

No, because pointers are of 32 bits :-)


> What would we have to accomplish in order to get a working x32? I'm assuming that ia32 and x32 aren't the same thing?

As you could have already seen, no, they are not the same thing ;-)


> @piranna are you referring to the x32 ABI? Maybe we should mention "ABI" for clarity, because I a bit confused at first, and others probably will be also.
> @lite20

Good idea, thanks by changuing the issue title :+1: 


> The whole thread is here: https://groups.google.com/forum/#!topic/v8-users/c-_URSZqTq8
> We can attempt to undertake the port as a third party and people may potentially join, or we could wait for the people using the platform to grow and be sufficient for the v8 team to do it themselves. As massive an undertaking as this might be, the performance gains shouldn't be overlooked.

Or we could simply just use the [correct arch option on Node.js](https://github.com/nodejs/node/blob/efebd0b79d930c50217caaddf1d245d510937bb0/configure#L32-L33) ;-)

**Comment by lite20** [comment]
> Yes, the same as in ia32/x86.
Once I found out you were reffering to the ABI it all made sense.

Gee, I know programs that will easily bust the 4 gigs of ram limit, especially on servers. Maybe we ought not deprecate the 64bit version to support said cases. If done right, three platforms shouldn't be too much to maintain?

**Comment by piranna** [comment]
No, they shouldn't only just to give x86_64 a little bit less love ;-)
Anyway, I find a single theaded process more than 4GB of data a little bit
bad designed... I vet it could be parallelized some way on several
simultaneous process and use several cores.

**Comment by lite20** [comment]
@piranna each thread can use 4GB? Mkay that makes a bit more sense. It's generally communication servers that will bust that 4GB mark.

**Comment by piranna** [comment]
Threads share the memory region, so no, apps using threads will still be
limited to 4GB.

**Comment by lite20** [comment]
@piranna just when I was beginning to understand hahaha. Okay, so that's what I had originally thought. What's your opinion of Minecraft servers than which will use 16gb RAM if it's available? :P

**Comment by piranna** [comment]
Probably they are not single threaded ;-) Is there any of them build in
node.js?

**Comment by lite20** [comment]
XD no, I don't think so, and I sure hope they aren't single threaded but I thought you said all threads of a program share the same heap?

**Comment by piranna** [comment]
Yes, al threads on a program share the heap, that's why they are usually
faster for intercomunication but also more insecure.

**Comment by lite20** [comment]
ah okay that makes sense @piranna. Comming from my high-level programming to low level programming always leaves a few knowledge gaps. Thanks for helping me out :stuck_out_tongue:

**Comment by piranna** [comment]
> ah okay that makes sense @piranna. Comming from my high-level programming to low level programming always leaves a few knowledge gaps. Thanks for helping me out :stuck_out_tongue:

You are welcome :-) I always loved to know the low level bits, so... :-P
